### DISCUSSION TITLE
Customizing the TextEdit widget

### DISCUSSION THREAD

**Discussion by mbillingr** [emilk/egui#1365] [Q&A] [answered]
Hi there!

I'm trying to create a widget for displaying and editing [S-expressions](https://en.wikipedia.org/wiki/S-expression).
Basically, my approach is to take an `egui::TextEdit` and plug in a custom `LayoutJob`, which works great for displaying.
However, for editing the `TextEdit`'s own cursor gets in the way. My editing logic has its own notion where the cursor should be, and this is usually not where the `TextEdit` puts it.

Is there a way to programmatically set cursor position and text selection in a `TextEdit` and/or control which input events it responds to?

Alternatively, just hiding the cursor would work for me too, but the only way I managed that was by setting `.interactive(false)`, which also seems to remove the useful ability to focus the widget.

Did I miss any customization hooks, or is there a better way for creating widgets?

**Comment by mbillingr** [comment]
Maybe I was misled by the lure of syntax highlighting and free text editing :smile: 

I'm now taking the much simpler approach of placing rich text labels according to my layout needs.

**Comment by DusterTheFirst** [comment, marked as answer]
Have you looked into the way the demo does syntax highlighting?

https://github.com/emilk/egui/blob/master/egui_demo_lib/src/apps/demo/code_editor.rs (https://www.egui.rs/#demo open code editor)

https://github.com/emilk/egui/blob/master/egui_demo_lib/src/easy_mark/easy_mark_editor.rs (https://www.egui.rs/#easymark)

**Reply by mbillingr** [reply]
Thanks for the links! I have indeed looked into the highlighting demos and they are super helpful.

I have not found any demo that shows how to exert programmatic control over the editor's cursor, though.... not sure if that's even possible.

**Reply by emilk** [reply]
It's possible! I've added an example in https://github.com/emilk/egui/commit/ccbddcfe951e01c55efd0ed19f2f2ab5edfad5d9 but you need to use latest `master`.

**Reply by mbillingr** [reply]
Wow, thanks! I wouldn't have figured that out on my own.
